accord-core/src/main/java/accord/coordinate/AbstractCoordinatePreAccept.java

```diff
@@ -0,0 +1,252 @@
+    {
+        Invariants.checkState(!initialRoundIsDone || (extraEpochs != null && !extraEpochs.extraRoundIsDone));
+        initialRoundIsDone = true;
+        if (extraEpochs != null)
+            extraEpochs.extraRoundIsDone = true;
+
+        if (failure instanceof CoordinationFailed)
+        {
+            ((CoordinationFailed) failure).set(txnId, route.homeKey());
+            if (failure instanceof Timeout)
+                node.agent().metricsEventsListener().onTimeout(txnId);
+            else if (failure instanceof Preempted)
+                node.agent().metricsEventsListener().onPreempted(txnId);
+            else if (failure instanceof Invalidated)
+                node.agent().metricsEventsListener().onInvalidated(txnId);
+        }
+        super.setFailure(failure);
+    }
+
+    final void onPreAcceptedOrNewEpoch()
+    {
+        Invariants.checkState(!initialRoundIsDone || (extraEpochs != null && !extraEpochs.extraRoundIsDone));
+        initialRoundIsDone = true;
+        if (extraEpochs != null)
+            extraEpochs.extraRoundIsDone = true;
+
+        // if the epoch we are accepting in is later, we *must* contact the later epoch for pre-accept, as this epoch
+        // could have moved ahead, and the timestamp we may propose may be stale.
+        // Note that these future epochs are non-voting, they only serve to inform the timestamp we decide
+        long latestEpoch = executeAtEpoch();
+        if (latestEpoch <= topologies.currentEpoch())
+            onPreAccepted(topologies);
+        else
+            onNewEpoch(topologies, latestEpoch);
+    }
+
+    final void onNewEpoch(Topologies prevTopologies, long latestEpoch)
+    {
+        // TODO (desired, efficiency): check if we have already have a valid quorum for the future epoch
+        //  (noting that nodes may have adopted new ranges, in which case they should be discounted, and quorums may have changed shape)
+        node.withEpoch(latestEpoch, () -> {
+            TopologyMismatch mismatch = TopologyMismatch.checkForMismatch(node.topology().globalForEpoch(latestEpoch), txnId, route.homeKey(), txn.keys());
+            if (mismatch != null)
+            {
+                initialRoundIsDone = true;
+                onNewEpochTopologyMismatch(mismatch);
+                return;
+            }
+            topologies = node.topology().withUnsyncedEpochs(route, txnId.epoch(), latestEpoch);
+            boolean equivalent = topologies.oldestEpoch() <= prevTopologies.currentEpoch();
+            for (long epoch = topologies.currentEpoch() ; equivalent && epoch > prevTopologies.currentEpoch() ; --epoch)
+                equivalent = topologies.forEpoch(epoch).shards().equals(prevTopologies.current().shards());
+
+            if (equivalent)
+            {
+                onPreAccepted(topologies);
+            }
+            else
+            {
+                extraEpochs = new ExtraEpochs(prevTopologies.currentEpoch() + 1, latestEpoch);
```

Comment:
Correct.

Reply:
Note that this can only happen under ownership changes, so the new owners would need to have agreed some newer transaction. I think in practice this is almost impossible, as the new epochs would have to have initiated an `ExclusiveSyncPoint` to be able to begin coordinating the ranges, so the `TxnId` we're coordinating would likely reject it - but it is possible the coordinator proposed a future `TxnId`.

At some point it will be worth elaborating a control flow diagram for all of this stuff.

Reply:
What do you mean the TxnId we are coordinating would likely reject it? 

Seems like for ephemeral reads we technically only have to do this once because we don't care about performing a stale read from Accord's perspective it just has to gather more dependencies if the nodes it originally contacted might not have known them all.